### ISSUE TITLE
Introducing dim_order: A New API for Understanding Dense Tensor Memory Layout

### ISSUE THREAD

**Issue by Gasoonjia** [pytorch/pytorch#146086]
As part of our ongoing efforts to improve the usability and performance of PyTorch, we are excited to introduce a new API called `dim_order`. This API provides a way to determine and return the physical layout of a dense tensor in memory, called dim order, which can help users optimize performance-critical code. Additionally, we provide functionality for users to raise expectations when ambiguous dim order occur.

### What is Dim Order?
In PyTorch, the dim order of a dense tensor represents the order in which its dimensions are laid out in memory. This is also known as the physical layout of the tensor. Dim order is represented as a tuple of integers, where each integer corresponds to a dimension in the tensor. Note that for tensors where memory format is defined (e.g. 4D tensor with dense memory layout) there is a one-to-one mapping between dim order and memory format.  For example, `torch.contiguous_format` for a 4D tensor would be represented as `(0, 1, 2, 3)`, indicating that the dimensions are stored in the order of batch, channel, height, and width. On the other hand, `torch.channels_last` would be represented as `(0, 2, 3, 1)`, indicating that the channel dimension is stored last.

Compared to `torch.memory_format`, `dim_order` provides an explicit and detailed representation of each dimension's meaning and is directly accessible as part of PyTorch's IR for dense tensor.

### Ambiguous Dim Order
In some cases, depending on the shape, a tensor may have multiple legal dim orders to describe its memory layout. We refer to such cases as dim order ambiguity. These cases are common in real-world scenarios. For instance, consider a tensor with a 4D size and a batch size equal to 1:

```
t1 = torch.zeros(1, a, b, c)  # a, b, and c are arbitrary integers 
```

Any of `(0, 1, 2, 3)`, `(1, 0, 2, 3)`, `(1, 2, 0, 3)`, and `(1, 2, 3, 0)` would be valid dim orders for `t1`. Therefore, we say that `t1` has an ambiguous dim order.

### Problems with Dim Order Ambiguity
The dimension order of each tensor should remain constant as long as the model graph remains unchanged. This means that the arrangement of data in memory for intermediate tensors should not be affected by the shape of the input tensors used for tracing, ensuring correctness and performance across different input shapes.

Let's revisit the previous example. When the batch size equals 1, any of the four tuples would be a valid dim order. However, when increasing the batch size to 2:

```
t1 = torch.zeros(2, a, b, c)  # a, b, and c are arbitrary integers
```

Now `t1` has only one valid dim order `(0, 1, 2, 3)`, which differs from the previous `t1`.

This can lead to an unexpected behavior where the memory layout becomes tensor shape sensitive when exporting with an example input. However, it is expected that the memory layout remains invariant to the shape of the tensor, meaning that changing the shape should not cause the memory layout to change. Therefore, we need to address the ambiguity issue.

Our proposed solution for this is to enhance the `torch.Tensor.dim_order()` API to allow users to detect ambiguity for a given shape.
### How to Avoid or Detect Ambiguity?
We introduced an extra argument called `ambiguity_check` to detect and resolve ambiguity in three ways:

- Set `ambiguity_check` to `False` (default) to return one of the legal dim orders without verifying its uniqueness.

- Set `ambiguity_check` to `True` to raise a `RuntimeError` if the dim order cannot be uniquely determined.

- Pass a list of `torch.memory_format` to check if the tensor conforms to one of the provided formats, raising a `RuntimeError` if no unique match is found.

To achieve anti-ambiguity within user control, we recommend passing all expected memory formats to `ambiguity_check`. For example, if the current model only uses `torch.contiguous_format` and `torch.channels_last` memory formats, we suggest using the following approach to obtain an unambiguous dim order:

```
allowed_formats = [torch.contiguous_format, torch.channels_last]
dim_order = tensor.dim_order(ambiguity_check=allowed_formats) # tensor here is an arbitrary pytorch tensor
```

This approach ensures that the obtained dim order is unambiguous within the current context.

### Example Usage
Here's an example of how to use the `dim_order` API:

```
import torch
# Create a tensor with shape (2, 3, 5, 7)
tensor = torch.empty((2, 3, 5, 7))


# A. Get the dim order of the tensor
dim_order = tensor.dim_order()
print(dim_order)  # Output: (0, 1, 2, 3)

# B. Check if the dim order is unique
try:
    tensor.dim_order(ambiguity_check=True)
except RuntimeError as e:
    print(e)  # Output: The tensor does not have unique dim order...

# C. Specify a list of allowed memory formats to resolve the ambiguity
allowed_formats = [torch.contiguous_format, torch.channels_last]
dim_order = tensor.dim_order(ambiguity_check=allowed_formats)
print(dim_order)  # Output: (0, 1, 2, 3)
```

Here’s the Colab link you can try: [dim_order_example.ipynb](https://colab.research.google.com/drive/1NBLW7k5sgz6mT8SpaphVxAo6w1qkjJol?usp=sharing)

By using the `dim_order` API, you can easily determine the physical layout of your tensors in memory and ensure that your models are optimized for performance.

### Best Practices
To get the most out of the `dim_order` API, we recommend the following best practices:
- Forward all allowed/expected memory formats to the `ambiguity_check` argument to return expected dim order and avoid ambiguity.
- If the expected memory format of the model is unclear, specify the `ambiguity_check` argument as `True` to ensure that ambiguities are detected.
- Verify that the returned dim order matches your expectations before using it in your model.

### Conclusion
In this post, we introduced the `dim_order` API, a new tool for understanding tensor memory layout in PyTorch. We showed how to use the API to determine the physical layout of tensors in memory and how to resolve ambiguity in dim order. We also provided best practices for using the API effectively. With the `dim_order` API, you can have a better understanding of current memory layout to optimize your models for performance and accuracy, leading to better results in your deep learning projects.

Great thanks @digantdesai,  @larryliu0820 and @ezyang for continued support and discussion!

**Comment by albanD** [comment]
I'm quite confused here, this API already exists?

But looking at the docs page, it is completely broken, so will repurpose this issue to tracking fixing the doc for this feature: https://pytorch.org/docs/stable/generated/torch.Tensor.dim_order.html

**Comment by digantdesai** [comment]
Yes, the API exists, and the new thing is an optional arg to that API. This post is trying to be more complete from dim_order point of view, I guess. @Gasoonjia - we should update the title to reflect this.

> But looking at the docs page, it is completely broken, so will repurpose this issue to tracking fixing the doc for this feature

Wow. Yeah, let's fix this ASAP.

**Comment by Gasoonjia** [comment]
Rename the title and rephase the start paragraphs to introduce our idea better. 
https://github.com/pytorch/pytorch/pull/146641 fixed the docs page
